x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plugin/TransportActionUtils.java

```diff
@@ -78,4 +78,56 @@ public static void executeRequestWithRetryAttempt(
             queryRunner.accept(onFailure);
         }
     }
+
+    public static void retryOnNodeWithMatchingVersion(
+        ClusterService clusterService,
+        QlVersionMismatchException exception,
+        Consumer<DiscoveryNode> retryRequest,
+        Consumer<Exception> onFailure,
+        Logger log
+    ) {
+        if (log.isDebugEnabled()) {
+            log.debug("Version mismatch detected: {}", exception.getMessage());
+        }
+
+        DiscoveryNode candidateNode = findNodeWithVersionOtherThanLocalNode(clusterService, exception.getInputVersion());
+
+        if (candidateNode != null) {
+            if (log.isDebugEnabled()) {
+                log.debug(
+                    "Candidate node to resend the request to: address [{}], id [{}], name [{}], version [{}]",
+                    candidateNode.getAddress(),
+                    candidateNode.getId(),
+                    candidateNode.getName(),
+                    candidateNode.getVersion()
+                );
+            }
+
+            retryRequest.accept(candidateNode);
+        } else {
+            if (log.isInfoEnabled()) {
+                log.info(
+                    "No candidate node with version {} found, likely all were upgraded in the meantime. Failing request.",
+                    exception.getInputVersion()
+                );
+            }
+            onFailure.accept(exception);
+        }
+
+    }
+
+    private static DiscoveryNode findNodeWithVersionOtherThanLocalNode(ClusterService clusterService, Version expectedVersion) {
+        DiscoveryNode localNode = clusterService.state().nodes().getLocalNode();
+        if (expectedVersion.equals(localNode.getVersion())) {
+            return null;
```

Comment:
Not sure I understand the logic here: if current node has the right version, it should be a good candidate itself... and a QlVersionMismatchException shouldn't have been thrown in the first place. Am I missing something?

Reply:
Yes, that's right. This condition should not happen and I don't think you can end up in this situation. But if it is the case, you would want to avoid the retry because you could end up in an endless loop otherwise. Throwing an exception here is probably even better.

Reply:
Right, it makes sense. 
The exception makes the intention much more clear 👍